### PR TITLE
MODINREACH-96 D2IR Local Endpoint (Circulation): Report Item Received to Owning Site for Item Hold

### PR DESCRIPTION
**Pull request by maryna-steshenko**
## Purpose

Add endpoint to receive message from central server to owning site indicating that a shipped item has been received by the borrowing site.
JIRA: [MODINREACH-96](https://issues.folio.org/browse/MODINREACH-96)
## Approach

- InnReachCirculationController modified.
- InnReachCirculationService modified.
- OpenAPI configuration modified.
- DTO schema added.
- InnReachCirculationController tests added.
## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/test/java/org/folio/innreach/controller/d2ir/InnReachCirculationControllerTest.java [hunk 1]
```diff
@@ -231,4 +233,56 @@ void processCancelItemHoldRequest_whenItemIsCheckedOut() {
...
+  void processItemReceivedRequest_whenItemIsShipped() {
+    var transaction = repository.findByTrackingIdAndCentralServerCode(PRE_POPULATED_TRACKING_ID,
+      PRE_POPULATED_CENTRAL_CODE).get();
+    transaction.setState(ITEM_SHIPPED);
+    repository.save(transaction);
+
+    var transactionHoldDTO = createTransactionHoldDTO();
+
+    var responseEntity = testRestTemplate.exchange(
+      "/inn-reach/d2ir/circ/itemreceived/{trackingId}/{centralCode}", HttpMethod.PUT,
```
Comment (oliinyko) [inline src/test/java/org/folio/innreach/controller/d2ir/InnReachCirculationControllerTest.java]: please add a constant for the endpoint path and reuse it in other tests

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java [hunk 2]
```diff
@@ -198,6 +201,19 @@ public InnReachResponseDTO cancelItemHold(String trackingId, String centralCode,
     return success();
   }
 
+  @Override
+  public InnReachResponseDTO itemReceived(String trackingId, String centralCode, ItemReceivedDTO itemReceivedDTO) {
+    var transaction = getTransaction(trackingId, centralCode);
+
+    if (transaction.getState() != ITEM_SHIPPED){
+      throw new IllegalArgumentException("Item is not shipped.");
+    }
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: lets unify the message for invalid transaction state. There are couple of examples in other methods:
`    Assert.isTrue(transaction.getState() == ITEM_SHIPPED,
      "Unexpected transaction state: " + transaction.getState());`

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java [hunk 3]
```diff
@@ -198,6 +201,19 @@ public InnReachResponseDTO cancelItemHold(String trackingId, String centralCode,
...
+    transaction.setState(ITEM_RECEIVED);
+    transactionRepository.save(transaction);
+
+    return success();
+  }
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: Do you know if there is any other story to handle the request payload? Should we validate the received data or update transaction details with it?

File: src/test/java/org/folio/innreach/controller/d2ir/InnReachCirculationControllerTest.java (hunk 1 above)
Comment (maryna-steshenko) [reply src/test/java/org/folio/innreach/controller/d2ir/InnReachCirculationControllerTest.java]: Do you mean add constant for the whole path or for the operation name only?

Comment (oliinyko) [reply src/test/java/org/folio/innreach/controller/d2ir/InnReachCirculationControllerTest.java]: The whole path "/inn-reach/d2ir/circ/itemreceived/{trackingId}/{centralCode}"

**Review by dmtkachenko — APPROVED** [review]

**Review by oliinyko — APPROVED** [review]

**Pull request merged**